src/Compiler/Postprocessor.php

```diff
@@ -30,26 +37,61 @@ final class Postprocessor
 
     private CompilerInterface $compiler;
 
+    private ScopeFactory $scopeFactory;
+
+    private NodeScopeResolver $nodeScopeResolver;
+
     public function __construct(
         Parser $parser,
         NodeVisitorStorage $nodeVisitorStorage,
         Standard $printerStandard,
-        CompilerInterface $compiler
+        CompilerInterface $compiler,
+        ScopeFactory $scopeFactory,
+        NodeScopeResolver $nodeScopeResolver
     ) {
         $this->parser = $parser;
         $this->nodeVisitorStorage = $nodeVisitorStorage;
         $this->printerStandard = $printerStandard;
         $this->compiler = $compiler;
+        $this->scopeFactory = $scopeFactory;
+        $this->nodeScopeResolver = $nodeScopeResolver;
     }
 
-    public function postProcess(string $phpContent, Template $template): string
+    public function postProcess(string $phpContent, Template $template, string $compileFilePath): string
     {
         $phpStmts = $this->findNodes($phpContent);
         foreach ($this->nodeVisitorStorage->getNodeVisitors() as $nodeVisitors) {
             $phpStmts = $this->processNodeVisitors($phpStmts, $nodeVisitors, $template);
         }
 
-        return $this->printerStandard->prettyPrintFile($phpStmts);
+        $phpContent = $this->printerStandard->prettyPrintFile($phpStmts);
+        file_put_contents($compileFilePath, $phpContent);
+        $realPath = realpath($compileFilePath) ?: '';
+        if ($realPath === '') {
+            return '';
+        }
+        require($compileFilePath); // load type definitions from compiled template
+
+        $scope = $this->scopeFactory->create(ScopeContext::create($realPath));
+
+        $phpStmts = $this->findNodes($phpContent);
+
+        $nodeScopeResolver = clone $this->nodeScopeResolver;
+        $nodeScopeResolver->processNodes($phpStmts, $scope, function (Node $node, Scope $scope) {
+            if ($node instanceof Expr && !$node instanceof EncapsedStringPart) {
+                $node->setAttribute(ExprTypeNodeVisitorInterface::TYPE_ATTRIBUTE, $scope->getType($node));
+            }
+        });
+
+        foreach ($this->nodeVisitorStorage->getNodeVisitors(true) as $nodeVisitors) {
+            $phpStmts = $this->processNodeVisitors($phpStmts, $nodeVisitors, $template);
+        }
+        $phpContent = $this->printerStandard->prettyPrintFile($phpStmts);
+
+        // update contents
+        file_put_contents($compileFilePath, $phpContent);
```

Comment:
If previous version of class is already required what would happed if it is rewritten here? I would think it should raise error "Class XY already declared"

Reply:
We are not requiring the same file twice. Here is just content update